doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,881 @@
+Presumably, this is possible trivially because it's already implemented as a
+WinRT type.
+
+I'll be especially curious to make sure that
+`TermControl::OnCreateAutomationPeer` can occur off the window process's UI
+thread, because all cross-process calls will need to happen off the UI thread.
+
+</td>
+</tr>
+<tr>
+<td><strong>Security</strong></td>
+<td>
+
+As we'll be introducing a mechanism for crossing elevation boundaries, we'll
+want to be especially careful as we make these changes.
+
+Our biggest concern regarding mixed elevation was regarding the ability for a
+non-elevated process to send input to another unelevated WT window, which was
+connected to an elevated client process. With these proposed changes, we won't
+have any unelevated windows connected to unelevated processes, so that concern
+should be mitigated.
+
+Additionally, we'll probably want to make sure that the user is visually
+prompted when a connection to an elevated client is initiated. Typically, this
+is done with a UAC prompt, which seems reasonable in this scenario as well.
+
+We'll likely want to default elevated windows to spawning unelevated
+connections, as to prevent accidentally running connections as an administrator.
+This is in contrast to existing behavior, where all connections in an elevated
+WT window are elevated.
+
+Furthermore, we'll want to ensure that there's nothing that an unelevated client
+process could do to trigger any sort of input callback in the parent window. If
+the parent window is an elevated window, with other elevated connections, then
+we don't want the uelevated content process to act as a vector by which
+malicious software could hijack the elevated window.
+
+</td>
+</tr>
+<tr>
+<td><strong>Reliability</strong></td>
+<td>
+
+This is probably the biggest concern in this section. Because there will now be
+many, many more processes in the process tree, it is imperitive that whenever
+we're doing cross-process operations, we do them safely.
+
+Whenever we're working with an object that's hosted by another process, we'll
+need to make sure that we work with it in a try/catch, because at _any_ time,
+the other process could be killed. At any point, a content process could be
+killed, without the window being closed. The window will need to be redundant to
+such a scenario, and if the content process is killed, display an appropriate
+error message, but _continue running_.
+
+When the monarch process dies, we'll need to be able to reliably elect a new
+monarch. While we're electing a new monarch, and updating the new monarch,
+there's always the chance that the _new_ monarch is also killed (This is the
+"Pope Stephen II" scenario). In this sccenario, if at any point a servant
+notices that the monarch has died, the servant will need to begin checking who
+the new monarch is again.
+
+There is certain to be a long tail of edge cases we'll discover as a product of

```

Comment:
Is there anything we can do to make tracking these bugs easier? I imagine a stack trace might be harder to get. Maybe we could have some added telemetry for this kind of stuff to be able to track down these root causes faster?

My concern is that we get _non-descriptive_ bug reports. So we'll know something is wrong, but won't have a clue where to start looking.

Reply:
I've added a note about adding extensive logging to the terminal